### ISSUE TITLE
lowering bug for nested closures with sparams

### ISSUE THREAD

**Issue by vtjnash** [JuliaLang/julia#16340]
the `tvars` list for a nested closure with reused `tvars` names incorrectly adds both the inner and outer `T` to `tvars`, even though the Lambda only expects only sparam.

```julia
julia> function f{T}(x::T)
           function g{T}(y::T)
             return (T,T)
           end
           return g
       end
f (generic function with 1 method)

julia> g=f(1)
g (generic function with 1 method)

julia> typeof(g).name.mt.defs.func.lambda_template.sparam_syms
svec(:T)

julia> typeof(g).name.mt.defs.tvars
svec(T,T)
```

this causes an assertion failure under julia-debug, and may be the source of https://github.com/dcjones/Gadfly.jl/issues/833#issuecomment-218505391

**Comment by protogeezer** [comment]
Is there any kind estimate of when this would be fixed?

**Comment by JeffBezanson** [comment]
I'll bump the priority on this. Will try my best to fix by 0.5-final.

**Comment by protogeezer** [comment]
I appreciate the activity on the problem. I am still seeing the same crash in Gadfly with the Julia sources pulled a few hours ago. 

```
et-imac-retina:julia-0.5-cxx sjbespa$ ./julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.0-dev+4891 (2016-06-21 21:43 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 5ed294d* (0 days old master)
|__/                   |  x86_64-apple-darwin15.6.0
```
